pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/relocation/RealtimeSegmentRelocationManager.java

```diff
@@ -0,0 +1,287 @@
+import com.linkedin.pinot.controller.helix.core.PinotHelixSegmentOnlineOfflineStateModelGenerator;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import org.apache.commons.collections.MapUtils;
+import org.apache.helix.HelixAdmin;
+import org.apache.helix.HelixManager;
+import org.apache.helix.model.IdealState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Manager to relocate completed segments from consuming servers to completed servers
+ * Segment relocation will be done by this manager, instead of directly moving segments to completed servers on completion,
+ * so that we don't get segment downtime when a segment is in transition
+ *
+ * We only relocate segments for realtime tables, and only if tenant config indicates that relocation is required
+ * A segment will be relocated, one replica at a time, once all of its replicas are in ONLINE state and on consuming servers
+ */
+public class RealtimeSegmentRelocationManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RealtimeSegmentRelocationManager.class);
+  private final ScheduledExecutorService _executorService;
+  private final PinotHelixResourceManager _pinotHelixResourceManager;
+  private final HelixManager _helixManager;
+  private final HelixAdmin _helixAdmin;
+  private final long _runFrequencyMinutes;
+
+  public RealtimeSegmentRelocationManager(PinotHelixResourceManager pinotHelixResourceManager, ControllerConf config) {
+    _pinotHelixResourceManager = pinotHelixResourceManager;
+    _helixManager = pinotHelixResourceManager.getHelixZkManager();
+    _helixAdmin = pinotHelixResourceManager.getHelixAdmin();
+    _runFrequencyMinutes = config.getRelocationManagerFrequencyInMinutes();
+
+    _executorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
+      @Override
+      public Thread newThread(Runnable runnable) {
+        Thread thread = new Thread(runnable);
+        thread.setName("PinotRelocationManagerExecutorService");
+        return thread;
+      }
+    });
+  }
+
+  public void start() {
+    LOGGER.info("Starting relocation manager");
+
+    _executorService.scheduleWithFixedDelay(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          runRelocation();
+        } catch (Exception e) {
+          LOGGER.warn("Caught exception while running relocation manager", e);
+        }
+      }
+    }, 5, _runFrequencyMinutes, TimeUnit.MINUTES);
+  }
+
+  public void stop() {
+    _executorService.shutdown();
+  }
+
+  /**
+   * Check all tables. Perform relocation of segments if table is realtime and relocation is required
+   */
+  public void runRelocation() {
+    if (!_pinotHelixResourceManager.isLeader()) {
```

Comment:
every back ground service needs to check for this. Is it possible to have something that manages background services that starts and stops on leadership change?

Reply:
I can look into this as part of a future PR, will avoid trying to touch so many files in this PR as it is only for RelocatorService